File: commander/src/main/java/com/iluwatar/commander/queue/Queue.java
/*
 * The MIT License
 * Copyright © 2014-2021 Ilkka Seppälä
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.iluwatar.commander.queue;

import com.iluwatar.commander.exceptions.IsEmptyException;

/**
 * Queue data structure implementation.
 *
 * @param <T> is the type of object the queue will hold.
 */
public class Queue<T> {

  private Node<T> front;
  private Node<T> rear;
  private int size;

  static class Node<V> {
    V value;
    Node<V> next;

    Node(V obj, Node<V> b) {
      value = obj;
      next = b;
    }
  }

  boolean isEmpty() {
    return size == 0;
  }

  void enqueue(T obj) {
    if (front == null) {
      front = new Node<>(obj, null);
      rear = front;
    } else {
      var temp = new Node<>(obj, null);
      rear.next = temp;
      rear = temp;
    }
    size++;
  }

  T dequeue() throws IsEmptyException {
    if (isEmpty()) {
      throw new IsEmptyException();
    } else {
      var temp = front;
      front = front.next;
      size = size - 1;
      return temp.value;
    }
  }

  T peek() throws IsEmptyException {
    if (isEmpty()) {
      throw new IsEmptyException();
    } else {
      return front.value;
    }
  }
}

File: commander/src/main/kotlin/com/iluwatar/commander/queue/Queue.kt
/*
 * This project is licensed under the MIT license. Module model-view-viewmodel is using ZK framework licensed under LGPL (see lgpl-3.0.txt).
 *
 * The MIT License
 * Copyright © 2014-2022 Ilkka Seppälä
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

// ABOUTME: Queue data structure implementation using linked nodes.
// ABOUTME: Provides enqueue, dequeue, peek, and isEmpty operations.
package com.iluwatar.commander.queue

import com.iluwatar.commander.exceptions.IsEmptyException

class Queue<T> {
    private var front: Node<T>? = null
    private var rear: Node<T>? = null
    private var size: Int = 0

    private class Node<V>(
        val value: V,
        var next: Node<V>? = null,
    )

    fun isEmpty(): Boolean = size == 0

    fun enqueue(obj: T) {
        if (front == null) {
            front = Node(obj)
            rear = front
        } else {
            val temp = Node(obj)
            rear?.next = temp
            rear = temp
        }
        size++
    }

    @Throws(IsEmptyException::class)
    fun dequeue(): T {
        if (isEmpty()) {
            throw IsEmptyException()
        }
        val temp = front
        front = front?.next
        size--
        return temp!!.value
    }

    @Throws(IsEmptyException::class)
    fun peek(): T {
        if (isEmpty()) {
            throw IsEmptyException()
        }
        return front!!.value
    }
}